src/star-utils.cpp

```diff
@@ -15,24 +18,155 @@ bool CatalogStarMagnitudeCompare(const CatalogStar &a, const CatalogStar &b) {
  * Remove dim stars from a catalog.
  * @param catalog Original catalog. Not mutated.
  * @param maxMagnitude Stars dimmer than this (greater numerical magnitude) are removed.
- * @param maxStars If there are greater than `maxStars` many stars left after filtering by `maxMagnitude`, only the `maxStars` brightest of them are kept.
+ * @param maxStars If there are greater than `maxStars` many stars left after filtering by
+ * `maxMagnitude`, only the `maxStars` brightest of them are kept.
  * @return Newly constructed catalog containing only the sufficiently bright stars.
  */
 Catalog NarrowCatalog(const Catalog &catalog, int maxMagnitude, int maxStars) {
     Catalog result;
     for (int i = 0; i < (int)catalog.size(); i++) {
+        // Higher magnitude implies the star is dimmer
         if (catalog[i].magnitude <= maxMagnitude) {
             result.push_back(catalog[i]);
         }
     }
+    // TODO: sort regardless?
+    std::sort(result.begin(), result.end(), CatalogStarMagnitudeCompare);
+
     if (maxStars < (int)result.size()) {
-        std::sort(result.begin(), result.end(), CatalogStarMagnitudeCompare);
+        // std::sort(result.begin(), result.end(), CatalogStarMagnitudeCompare);
         result.resize(maxStars);
     }
 
     return result;
 }
 
+int KeyToIndex(std::vector<int> key, int binFactor, long long maxIndex) {
+    const long long MAGIC_RAND = 2654435761;
+    long index = 0;
+    for (int i = 0; i < (int)key.size(); i++) {
+        index += key[i] * std::pow(binFactor, i);
+    }
+
+    return ((index % maxIndex) * (MAGIC_RAND % maxIndex)) % maxIndex;
+}
+
+std::pair<std::vector<short>, std::vector<short>> TetraPreparePattCat(const Catalog &catalog,
```

Comment:
Could it be helpful to sort by magnitude before calling tetrapreparepattcat?  Right now it takes the first 10 stars in an fov for example, but what if the first 10 all happen to be very dim? May not be necessary though if it's already working alright.

Reply:
Should be ok, but fixed I think (with one of your commits)